File: 64.minimum-path-sum/main.go
package main

import "math"

func main() {
}

func minPathSum(grid [][]int) int {
	for i := 0; i < len(grid); i++ {
		for j := 0; j < len(grid[0]); j++ {
			if i == 0 && j == 0 {
				continue
			}
			left, up := math.MaxInt64, math.MaxInt64
			if i > 0 {
				up = grid[i-1][j]
			}
			if j > 0 {
				left = grid[i][j-1]
			}
			grid[i][j] += mn(left, up)
		}
	}
	return grid[len(grid)-1][len(grid[0])-1]
}

func mn(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: 64.minimum-path-sum/main.rs
use std::cmp::min;

impl Solution {
    pub fn min_path_sum(grid: Vec<Vec<i32>>) -> i32 {
        let mut m = vec![vec![0; grid[0].len()]; grid.len()];
        for i in 0..grid.len() {
            for j in 0..grid[0].len() {
                if i == 0 && j == 0 {
                    m[i][j] = grid[i][j];
                    continue;
                }
                if i == 0 {
                    m[i][j] = m[i][j-1] + grid[i][j];
                } else if j == 0{
                    m[i][j] = m[i-1][j] + grid[i][j];
                } else {
                    m[i][j] = min(m[i][j-1], m[i-1][j]) + grid[i][j];
                }
            }
        }
        m[grid.len()-1][grid[0].len()-1]
    }
}
